crates/edr_provider/src/pending.rs

```diff
@@ -0,0 +1,205 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use edr_eth::{receipt::BlockReceipt, SpecId, B256, U256};
+use edr_evm::{
+    blockchain::{Blockchain, BlockchainError, BlockchainMut, SyncBlockchain},
+    db::BlockHashRef,
+    state::{StateDiff, StateError, StateOverride, SyncState},
+    LocalBlock, SyncBlock,
+};
+
+/// A blockchain with a pending block.
+///
+/// # Panics
+///
+/// Panics if a state override is provided to `state_at_block_number` for the
+/// pending block; or if the `BlockchainMut` methods are called.
+#[derive(Debug)]
+pub(crate) struct BlockchainWithPending<'blockchain> {
```

Comment:
I'm a little bit concerned about the various asserts and panics and I'm wondering if we really need this type?

As far as I can tell, we only need `BlockchainWithPending` for the `dry_run` in `run_call` and `dry_run` only uses `spec_at_block_number` and `BlockHashRef`. So an alternative option could be to introduce an enum that is either a `SyncBlockchain` or a `SyncBlockchain` + a pending block, implement `spec_at_block_number` and `BlockHashRef` for this enum and have `dry_run` take this enum as argument instead of `SyncBlock`.

Reply:
I agree. Following your feedback, I rewrote the code such that the `runtime.rs` run-methods all only need the `BlockHashRef` trait, removing the need for the other implementations on `BlockchainWithPending`.

This is where I ran into trouble. We need to _upcast_ a dynamic trait object, which is currently not natively supported in Rust. Rust v1.76.0 will introduce this and is scheduled for release on February 8th, 2024.

To still fulfil your request, I tried all recommended workarounds for this issue that I could find, but none of them worked. After spending several hours on this, I feel that it's not worth spending more time on until native dynamic trait object upcasting is supported in Rust. I created [this issue](https://github.com/NomicFoundation/edr/issues/244) for it.

I believe that the documentation on `BlockchainWithPending` and is locality (`pub(crate)` should be sufficient to avoid usage with the unsupported trait implementations.